File: Task/Extra-primes/Ruby/extra-primes.rb
def nextPrimeDigitNumber(n)
    if n == 0 then
        return 2
    end
    if n % 10 == 2 then
        return n + 1
    end
    if n % 10 == 3 or n % 10 == 5 then
        return n + 2
    end
    return 2 + nextPrimeDigitNumber((n / 10).floor) * 10
end

def isPrime(n)
    if n < 2 then
        return false
    end
    if n % 2 == 0 then
        return n == 2
    end
    if n % 3 == 0 then
        return n == 3
    end
    if n % 5 == 0 then
        return n == 5
    end

    wheel = [4, 2, 4, 2, 4, 6, 2, 6]
    p = 7
    loop do
        for w in wheel
            if p * p > n then
                return true
            end
            if n % p == 0 then
                return false
            end
            p = p + w
        end
    end
end

def digitSum(n)
    sum = 0
    while n > 0
        sum = sum + n % 10
        n = (n / 10).floor
    end
    return sum
end

LIMIT = 10000
p = 0
n = 0

print "Extra primes under %d:\n" % [LIMIT]
while p < LIMIT
    p = nextPrimeDigitNumber(p)
    if isPrime(p) and isPrime(digitSum(p)) then
        n = n + 1
        print "%2d: %d\n" % [n, p]
    end
end
print "\n"

File: Task/Extra-primes/Rust/extra-primes.rs
// [dependencies]
// primal = "0.3"

fn is_prime(n: u64) -> bool {
    primal::is_prime(n)
}

fn next_prime_digit_number(n: u64) -> u64 {
    if n == 0 {
        return 2;
    }
    match n % 10 {
        2 => n + 1,
        3 | 5 => n + 2,
        _ => 2 + next_prime_digit_number(n / 10) * 10,
    }
}

fn digit_sum(mut n: u64) -> u64 {
    let mut sum = 0;
    while n > 0 {
        sum += n % 10;
        n /= 10;
    }
    return sum;
}

fn main() {
    let limit1 = 10000;
    let limit2 = 1000000000;
    let last = 10;
    let mut p = 0;
    let mut n = 0;
    let mut extra_primes = vec![0; last];
    println!("Extra primes under {}:", limit1);
    loop {
        p = next_prime_digit_number(p);
        if p >= limit2 {
            break;
        }
        if is_prime(digit_sum(p)) && is_prime(p) {
            n += 1;
            if p < limit1 {
                println!("{:2}: {}", n, p);
            }
            extra_primes[n % last] = p;
        }
    }
    println!("\nLast {} extra primes under {}:", last, limit2);
    let mut i = last;
    while i > 0 {
        i -= 1;
        println!("{}: {}", n - i, extra_primes[(n - i) % last]);
    }
}
